### PR TITLE
#167190489 Implement User Signup

### PR DESCRIPTION
**Pull request by abdulfatai360**
#### What does this PR do?
- Enable users to signup/register for an Authors Haven's account

#### Description of Task proposed in this pull request?
- Validate registration inputs users submit
- Notify users of any validation or signup errors
- Redirect users to profile page if signup request is successful

#### How should this be manually tested (Quality Assurance)?
- Run `npm run start:dev`
- In the browser that would be launched, enter `http://localhost:8080/signup` in the address bar
- Use the signup form and hit the submit button

#### What are the relevant pivotal tracker stories?
- [#167190489](https://www.pivotaltracker.com/story/show/167190489)

#### Any background context you want to add (Operations Impact)?
- Run `npm install` to download dependencies
- I created a dummy `Profile page` users can be redirected to when they successfully signed in. This page is solely created to test my signup implementation.
- I also created some reusable components (`Form`, `Input`, and `Button`) in this PR. These components had been created in another PR. I mocked them here so that it would be easier to couple the other PR with this PR easily. My file changes are much because of these components.
- For the profile page components and the reusable components, I didn't write tests for them

#### What I have learned working on this feature: 
- I gained a better understanding of React/Redux testing

#### Screenshots:
- The signup page (with validation error in one of the inputs)
signup_first_iteration

### CODE REVIEW

**Review by henryade — APPROVED** [review]

File: src/components/Signup/Signup.scss [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+.Signup__form {
+  width: 100%;
+  max-width: 400px;
+
+  .signup-error {
+    // border: 3px solid #dd2c00;
+    color: #dd2c00;
+    // background: #dd2c00;
+    // padding: 1rem 0.5rem;
```
Comment (ayodejiAA) [inline src/components/Signup/Signup.scss]: Why do we have these commented lines here? Please remove them.

**Review by ayodejiAA — APPROVED** [review]

**Review by henryade — APPROVED** [review]

File: src/components/Signup/Signup.scss (hunk 1 above)
Comment (abdulfatai360) [reply src/components/Signup/Signup.scss]: Thanks for pointing it out.

**Review by Adekoreday — APPROVED** [review]

File: src/reducers/user/user.spec.js [hunk 2]
```diff
@@ -2,6 +2,10 @@ import { GET_USER_PROFILE } from '../../actionTypes';
 import userReducer from './';
 
 describe('User Reducer', () => {
+  it('returns default state when state parameter is not passed', () => {
+    expect(userReducer(undefined, {})).toEqual(null);
```
Comment (dinobi) [inline src/reducers/user/user.spec.js]: Why is the default state null?

File: src/actions/user/signupAction.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import httpClient from '../../helpers/index';
+import {
+  USER_SIGNUP_REQUEST,
```
Comment (dinobi) [inline src/actions/user/signupAction.js]: `SIGNUP_*` is sufficient. Pls lets stick with a convention.

File: src/actions/user/signupAction.spec.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import configureStore from 'redux-mock-store';
+import moxios from 'moxios';
+import signupAction from './signupAction';
+import {
+  USER_SIGNUP_REQUEST,
+  USER_SIGNUP_SUCCESS,
+  USER_SIGNUP_FAILURE,
+} from '../../actionTypes';
+
+const MOCK_USER = {
```
Comment (dinobi) [inline src/actions/user/signupAction.spec.js]: can create a single source for our mocked resources, lets make our app DRY?

File: src/components/Input/Input.spec.js [hunk 5]
```diff
@@ -25,6 +25,13 @@ describe('Input Component', () => {
     expect(wrapper.find('input')).toHaveLength(1);
   });
 
+  it('renders an error below when it recieves error props', () => {
+    const { wrapper } = setup({ error: 'first name is required' });
+
+    expect(wrapper.exists()).toBeTruthy();
+    expect(wrapper.find('.input-validation-error')).toHaveLength(1);
```
Comment (dinobi) [inline src/components/Input/Input.spec.js]: you're supposed to initially check that `input-validation-error` is not present before the error, also check that `input-validation-error` contains error that occurred

File: src/components/Signup/Signup.jsx [hunk 6]
```diff
@@ -0,0 +1,255 @@
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import { Link } from 'react-router-dom';
+import Joi from 'joi-browser';
+import PropTypes from 'prop-types';
+import signupInputSchema from '../../schemas/signup';
+import signupAction from '../../actions/user/signupAction';
+import Form from '../Form/Form';
```
Comment (dinobi) [inline src/components/Signup/Signup.jsx]: can you create an index file for all our reusable components and import from there

File: src/components/Signup/Signup.jsx [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+import Input from '../Input/Input';
+import Button from '../Button/Button';
+import GoogleIcon from '../../../assets/icon/google.png';
```
Comment (dinobi) [inline src/components/Signup/Signup.jsx]: pls use an index file for this, consult @LukasChiama for this

File: src/components/Signup/Signup.jsx [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+      confirmPassword: '',
+    },
+    errors: {},
+  };
+
+  getValidationErrorMessages = (error) => {
+    const validationErrors = {};
+    const errorDetails = error.details;
+
+    for (let i = errorDetails.length - 1; i >= 0; i -= 1) {
```
Comment (dinobi) [inline src/components/Signup/Signup.jsx]: each time the loop runs `errorDetails.length` will be evaluated, why not get the size first before the loop starts.

File: src/components/Signup/Signup.jsx [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+  }
+}
+
+Signup.propTypes = {
+  signupReducer: PropTypes.object.isRequired,
+  signupAction: PropTypes.func.isRequired,
+  history: PropTypes.object.isRequired,
+};
+
+const mapStateToProps = ({ signupReducer }) => ({ signupReducer });
```
Comment (dinobi) [inline src/components/Signup/Signup.jsx]: why is this called signupReducer?, is it a function that takes in an action, a state and returns a new state?

File: src/actions/user/signupAction.js (hunk 3 above)
Comment (abdulfatai360) [reply src/actions/user/signupAction.js]: @dinobi I have resolved this.

File: src/actions/user/signupAction.spec.js (hunk 4 above)
Comment (abdulfatai360) [reply src/actions/user/signupAction.spec.js]: @dinobi I have created a `mockData.js` file in the `helpers/` directory. I did it this way to conform with what Adeola already implemented in his PR.

File: src/components/Input/Input.spec.js (hunk 5 above)
Comment (abdulfatai360) [reply src/components/Input/Input.spec.js]: @dinobi I have created a new test case to check that that `input-validation-error` is not present where there is nor error prop passed.

I have also asserted that the text content of `input-validation-error` is the same as the value of the error prop that was passed.

File: src/components/Signup/Signup.jsx [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+   *
+   * @returns
+   * @memberof Signup
+   */
+  render() {
+    const { data, errors } = this.state;
+    const { signupReducer } = this.props;
+
+    return (
+      <>
```
Comment (dinobi) [inline src/components/Signup/Signup.jsx]: Any reason why we have this?

File: src/components/Signup/Signup.jsx [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+        <main className="Signup">
+          <section className="Signup__info">
```
Comment (dinobi) [inline src/components/Signup/Signup.jsx]: There's a possibility we might need to abstract this into it's own component, as I've noticed that many other pages have this

File: src/components/Signup/Signup.jsx [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+                <Button
+                  className="btn"
+                  title="Please fill up the form"
+                  label={signupReducer.isLoading ? 'Please wait...' : 'Submit'}
+                  disabled={signupReducer.isLoading}
+                />
+              </div>
+            </Form>
+
+            <div className="Signup__form__footer">
```
Comment (dinobi) [inline src/components/Signup/Signup.jsx]: We can abstract this too. You can sync with Aluka about this

File: src/components/Signup/Signup.spec.js [hunk 13]
```diff
@@ -0,0 +1,244 @@
+import CheckPropTypes from 'check-prop-types';
+import { Signup } from './Signup';
+import Form from '../Form/Form';
+import Input from '../Input/Input';
+import Button from '../Button/Button';
+
+const MOCK_USER = {
```
Comment (dinobi) [inline src/components/Signup/Signup.spec.js]: why is this still here

File: src/components/Signup/Signup.jsx (hunk 9 above)
Comment (abdulfatai360) [reply src/components/Signup/Signup.jsx]: I have changed the name to `signup` as agreed during the TTL interaction

File: src/reducers/user/user.spec.js (hunk 2 above)
Comment (abdulfatai360) [reply src/reducers/user/user.spec.js]: @dinobi I have removed this implementation. It now resides in Aderokede's PR as discussed during the TTL interaction that the Header component should be the one getting user data.

File: src/components/Signup/Signup.jsx (hunk 8 above)
Comment (abdulfatai360) [reply src/components/Signup/Signup.jsx]: I have implemented this feedback too, @dinobi

File: src/components/Signup/Signup.jsx (hunk 6 above)
Comment (abdulfatai360) [reply src/components/Signup/Signup.jsx]: This has been handled in Juwon's PR. However, I added it in my PR in the case my PR gets merged before his.

File: src/components/Signup/Signup.jsx (hunk 7 above)
Comment (abdulfatai360) [reply src/components/Signup/Signup.jsx]: I consulted Chukwualuka and implemented the necessary changes.

File: src/components/Header/Header.jsx [hunk 14]
```diff
@@ -21,7 +20,7 @@ export class Header extends Component {
    * @memberof Header
    */
   componentDidMount() {
-    this.props.getUserProfile();
+    // this.props.getUserProfile();
```
Comment (dinobi) [inline src/components/Header/Header.jsx]: why commented code, if we don't need this kindly remove this method

**Pull request merged**